magma/circuit.py

```diff
@@ -204,14 +205,78 @@ def _get_intermediate_values(value):
     return values
 
 
+class NamerDict(dict):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._inferred_names = {}
+
+    def _set_name(self, key, value):
+        if isinstance(value, (Type, MagmaProtocol)):
+            key = TempNamedRef(key, value)
+        value.name = key
+
+    def _check_unique_name(
+        self,
+        key: str,
+        value: Union[Type, 'Circuit'],
+    ):
+        """If key has been seen more than once, "uniquify" the names by append
+           _{i} to them."""
+        values = self._inferred_names.setdefault(key, [])
+        if len(values) == 0:
+            self._set_name(key, value)
+        elif any(value is x for x in values):
+            # Already uniquified
+            return
+        else:
+            if len(values) == 1:
+                # Make the first value consistent by append _0
+                self._set_name(f"{key}_0", values[0])
+            self._set_name(f"{key}_{len(values)}", value)
+        values.append(value)
+
+    def _set_value_name(self, key: str, value: Type):
+        if not hasattr(value, "name"):
+            return  # Interface object is a Type without a name.
+        if isinstance(value.name, AnonRef):
+            self._check_unique_name(key, value)
+        elif isinstance(value.name, TempNamedRef):
+            self._check_unique_name(value.name.name, value)
+
+    def _set_inst_name(self, key: str, value: 'Circuit'):
+        if not value.name:
+            self._check_unique_name(key, value)
+        else:
+            self._check_unique_name(value.name, value)
+
+    def __setitem__(self, key, value):
+        super().__setitem__(key, value)
+        if isinstance(value, (Type, MagmaProtocol)):
+            self._set_value_name(key, value)
+        elif isinstance(type(value), CircuitKind):
+            # NOTE: we check type(value) because this code is run in the Circuit
+            # class creation pipeline (so Circuit may not be defined yet).
```

Comment:
```suggestion
            # NOTE(leonardt): we check type(value) because this code is run in the Circuit
            # class creation pipeline (so Circuit may not be defined yet).
```